addons/web/models/models.py

```diff
@@ -831,6 +835,289 @@ def group_id_name(value):
 
             return { 'values': field_range, }
 
+    def onchange2(self, values: Dict, field_name: str, fields_spec: Dict):
+        """ Perform an onchange on the given field.
+
+            :param values: dictionary mapping field names to values on the form view,
+                giving the current state of modification
+            :param field_name: name of the modified field, or list of field
+                names (in view order), or False
+            :param fields_spec: dictionary specifying the fields in the view,
+                just like the one used by :meth:`web_read`
+
+            When ``field_name`` is falsy, the method first adds default values
+            to ``values``, computes the remaining fields, applies onchange
+            methods to them, and return all the fields in ``fields_spec``.
+        """
+        # this is for tests using `Form`
+        self.env.flush_all()
+
+        env = self.env
+        if isinstance(field_name, list):
+            names = field_name
+        elif field_name:
+            names = [field_name]
+        else:
+            names = []
+
+        first_call = not names
+
+        if any(name not in self._fields for name in names):
+            return {}
+
+        class Snapshot(dict):
```

Comment:
Wouldn't it make sense to move this class to the module scope? A `class RecordSnapshot`, perhaps? The separation of concerns is clear: it's a tool to compare records and create diffs for a list of fields.

In my opinion, this method has already too many lines to hold a class definition. Also, such a class may possibly come in handy in other situations, if available.